File: objectivec/day6_part1_2021.m
#import <Foundation/Foundation.h>

int main() {
    @autoreleasepool {
        NSError *error = nil;
        NSString *input = [NSString stringWithContentsOfFile:@"input.txt" encoding:NSUTF8StringEncoding error:&error];
        NSArray *fishAges = [input componentsSeparatedByString:@","];
        
        NSMutableArray *currentFishAges = [NSMutableArray arrayWithArray:fishAges];
        
        for (int day = 1; day <= 80; day++) {
            NSMutableArray *newFishAges = [NSMutableArray array];
            
            for (NSString *age in currentFishAges) {
                NSInteger currentAge = [age integerValue];
                if (currentAge == 0) {
                    [newFishAges addObject:@"6"];
                    [newFishAges addObject:@"8"];
                } else {
                    [newFishAges addObject:[NSString stringWithFormat:@"%ld", currentAge - 1]];
                }
            }
            
            currentFishAges = newFishAges;
        }
        
        printf("Total number of lanternfish after 80 days: %lu\n", (unsigned long)currentFishAges.count);
    }
    
    return 0;
}

File: swift/day6_part1_2021.swift

import Foundation

let input = try String(contentsOfFile: "input.txt").trimmingCharacters(in: .whitespacesAndNewlines)
let lanternfishAges = input.components(separatedBy: ",").compactMap { Int($0) }

var currentAges = lanternfishAges
for day in 1...80 {
    var newAges: [Int] = []
    for age in currentAges {
        if age == 0 {
            newAges.append(6)
            newAges.append(8)
        } else {
            newAges.append(age - 1)
        }
    }
    currentAges = newAges
}

print(currentAges.count)
